File: Queue/queue.ts
let head: NodeElement | null = null;

export class NodeElement {
  pointer: NodeElement | null = null;
  data: any | null = null;

  constructor(data: any) {
    this.data = data;
  }
}

export function enQueue(element: NodeElement) {
  let headElement = head;
  if (!headElement) {
    head = element;
    return head;
  }
  while (headElement.pointer !== null) {
    headElement = headElement.pointer;
  }

  headElement.pointer = element;
}

export function deQueue() {
  if (head) {
    const oldElement = head;
    head = head.pointer;
    return oldElement;
  }
  return null;
}

export function QHasElement() {
  if (!head) {
    return false;
  }
  return true;
}

function printQueue() {
  console.log("Printing queue...");
  let headElement = head;
  while (headElement !== null) {
    console.log(headElement.data, "->");
    headElement = headElement.pointer!;
  }
}

// starting point
function start() {
  //add elements to queue
  enQueue(new NodeElement("1"));
  enQueue(new NodeElement("2"));
  enQueue(new NodeElement("3"));
  printQueue(); // print linked list
  deQueue();
  printQueue(); // print linked list
}

// start();

File: Golang/Queue/queue.go
package main

import (
	"fmt"
	"log"
)

type ListNode struct {
	Next *ListNode
	Data any
}

var HeadNode *ListNode
var LastNode *ListNode

func Enqueue(data string) *ListNode {
	newNode := &ListNode{
		Data: data,
	}

	if HeadNode == nil && LastNode == nil {
		HeadNode = newNode
		LastNode = newNode
		return newNode
	}

	(*LastNode).Next = newNode
	LastNode = newNode
	return newNode
}

func Dequeue() *ListNode {
	deleteNode := HeadNode

	if HeadNode == nil {
		log.Fatal("queue undeflow...")
	}

	HeadNode = HeadNode.Next
	return deleteNode
}

func PrintQueue() {
	fmt.Println("Printing Queue....")
	for listNode := HeadNode; listNode != nil; {
		fmt.Println(listNode.Data, "-->")
		listNode = listNode.Next
	}
}

func main() {
	Enqueue("1")
	Enqueue("2")
	Enqueue("3")
	Enqueue("4")

	PrintQueue()

	Dequeue()
	Dequeue()
	Dequeue()
	Dequeue()

	PrintQueue()

	Dequeue()
}
